src/dparse/parser.d

```diff
@@ -3556,6 +3593,128 @@ class Parser
         return node;
     }
 
+    /**
+     * Parses an AsmInstruction using GCC Assembler
+     *
+     * $(GRAMMAR $(RULEDEF gccAsmInstruction):
+     *     | $(RULE expression) $(LITERAL ':') $(RULE gccAsmOperandList) ($(LITERAL ':') $(RULE gccAsmOperandList) ($(LITERAL ':') $(RULE stringLiteralList))? )? $(LITERAL ';')
+     *     | $(RULE expression) $(LITERAL ':') $(LITERAL ':') $(RULE gccAsmOperandList) $(LITERAL ':') $(RULE stringLiteralList) $(LITERAL ';') $(LITERAL ':') $(RULE declaratorIdentifierList) $(LITERAL ';')
+     *     ;)
+     */
+    /*
+     * References:
+     * - [1] https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
+     * - [2] https://wiki.dlang.org/Using_GDC
+     * - [3] https://github.com/dlang/dmd/blob/master/src/dmd/iasmgcc.d
+     *
+     * Separated into a different method because one cannot interleave DMD & GCC asm
+     * <asm-qualifiers> (volatile, inline, goto) not supperted (yet?)
+     */
+    GccAsmInstruction parseGccAsmInstruction()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+        const startIndex = index;
+        auto node = allocator.make!GccAsmInstruction();
+
+        // Allow empty asm instructions
+        if (currentIs(tok!";"))
+        {
+            warn("Empty asm instruction");
+            node.tokens = tokens[startIndex .. index];
+            return node;
+        }
+
+        mixin(parseNodeQ!("node.assemblerTemplate", "Expression"));
+
+        // GDC allows e.g. asm { mixin(<some asm instruction>); }
+        if (!currentIs(tok!";"))
+        {
+            mixin(tokenCheck!":");
+            mixin(parseNodeQ!(`node.outputOperands`, `GccAsmOperandList`));
+
+            if (skip(tok!":"))
+            {
+                mixin(parseNodeQ!(`node.inputOperands`, `GccAsmOperandList`));
+
+                if (skip(tok!":"))
+                {
+                    if (!currentIs(tok!":"))
+                        mixin(parseNodeQ!("node.registers", "StringLiteralList"));
+
+                    if (skip(tok!":"))
+                    {
+                        if (node.outputOperands.items.length)
+                            error("goto-labels only allowed without output operands!", false);
```

Comment:
I think there may be a `return null;` missing after this `error`.

Reply:
Parsing the goto's anyway achieves better error reporting because it avoids parsing the remaining goto's as a new instruction.
But you're right, it should probably return null after that.